papaya-web/src/main/java/pl/edu/agh/papaya/rest/projects/controller/ProjectsController.java

```diff
@@ -98,4 +108,72 @@ private Project getValidProject(Long id) {
 
         return ResponseEntity.ok(sprints);
     }
+
+    @Override
+    public ResponseEntity<UserAvailabilityDto> getUserAvailability(Long projectId, Long sprintId) {
+        Project project = getValidProject(projectId);
+        Sprint sprint = getValidSprint(sprintId);
+
+        if (!isSprintInProject(sprint, project)) {
+            throw new ForbiddenAccessException();
+        }
+
+        Availability availability = availabilityService
+                .findBySprintIdAndUserId(sprintId, userContext.getUserIdn())
+                .orElseThrow(ResourceNotFoundException::new);
+
+        return ResponseEntity.ok(availabilityMapper.mapToApi(availability));
+    }
+
+    @Override
+    public ResponseEntity<UserAvailabilityDto> updateUserAvailability(@Valid UserAvailabilityDto body, Long projectId,
+            Long sprintId) {
+        Project project = getValidProject(projectId);
+        Sprint sprint = getValidSprint(sprintId);
+        UserInProject userInProject = getValidUserInProject(projectId);
+        if (!isSprintInProject(sprint, project) || !isAvailabilityDeclarationAllowed(sprint)) {
+            throw new ForbiddenAccessException();
+        }
+
+        Availability availability = availabilityService.findBySprintIdAndUserId(sprintId, userContext.getUserIdn())
+                .map(persistedAvailability -> updatePersistedAvailability(persistedAvailability, body))
+                .orElseGet(() -> createAvailability(body, sprint, userInProject));
+
+        return ResponseEntity.ok(availabilityMapper.mapToApi(availability));
+    }
+
+    private Sprint getValidSprint(Long sprintId) {
+        return sprintService.findById(sprintId)
+                .orElseThrow(ResourceNotFoundException::new);
+    }
+
+    private UserInProject getValidUserInProject(Long projectId) {
+        return userInProjectService.findByProjectIdAndUserId(projectId, userContext.getUserIdn())
+                .orElseThrow(ResourceNotFoundException::new);
+    }
+
+    private boolean isSprintInProject(Sprint sprint, Project project) {
+        return sprint.getProject().getId().equals(project.getId());
+    }
+
+    private boolean isAvailabilityDeclarationAllowed(Sprint sprint) {
+        return sprint.getSprintState() == SprintState.DECLARABLE;
+    }
+
+    private Availability createAvailability(UserAvailabilityDto body, Sprint sprint, UserInProject userInProject) {
+        return availabilityService.newAvailability()
+                .withSprint(sprint)
+                .withUserInProject(userInProject)
```

Comment:
Nie powinniśmy tutaj podać samego usera?

Reply:
Availability ma pole UserInProject, więc i tak musiałoby to być gdzieś wyciągnięte z bazy (a tutaj akurat już to jest)

Reply:
"Akurat" już jest bo wyciągnąłeś ;) co jeśli ktoś poda `userInProject` z projektem, który nie ma sprintu `sprint`? Powinniśmy raczej dawać minimalną, lecz wystarczającą ilość informacji.

Reply:
Równie dobrze ktoś może podać usera, który nie jest w projekcie, w którym jest dany sprint.
Moim zdaniem to do osoby wykorzystującej kreator należy upewnienie się, czy podawane dane mają sens

Reply:
> Równie dobrze ktoś może podać usera, który nie jest w projekcie, w którym jest dany sprint.

Różnica jest taka, że informacja o userze jest niezbędna, a o projekcie jest zbędna.

> Moim zdaniem to do osoby wykorzystującej kreator należy upewnienie się, czy podawane dane mają sens

Noo ciekawe zdanie, nie powiem... Czyli jak wywołujemy metodę to poprawność argumentów sprawdzamy przed każdym wywołaniem a nie w metodzie?

Reply:
Poprawność danych sprawdzam w metodzie, ale kontrolera. Wszystko co jest pod spodem nie musi nic sprawdzać, bo jako osoba wykorzystująca wizarda zadbałem o to, aby dane miały sens. 

Sugerujesz, żeby w wizardzie sprawdzać, czy podany User jest w projekcie, w którym jest podany Sprint?

Reply:
Btw. odgapiłem implementację od Ciebie i w przypadku podania np. nulla leci wyjątek z treścią błędu, która jest wysyłana na front

Reply:
No i warto dodać, że treść błędu to coś w stylu "pole xxx nie może być nullem", więc szczegóły implementacji lecą na front. 
W takim przypadku dodaje się jakieś mappery, które przerabiają ten null pointer na jakiś error http?

Reply:
> Poprawność danych sprawdzam w metodzie, ale kontrolera. Wszystko co jest pod spodem nie musi nic sprawdzać, bo jako osoba wykorzystująca wizarda zadbałem o to, aby dane miały sens.

Chodziło mi raczej o:
_Czyli jak wywołujemy metodę X to poprawność argumentów sprawdzamy przed każdym wywołaniem metody X a nie w metodzie X?_

> Sugerujesz, żeby w wizardzie sprawdzać, czy podany User jest w projekcie, w którym jest podany Sprint?

Tak, tak samo jak sprawdzamy czy coś jest nullem oraz czy user ma uprawnienia. Te rzeczy powinny być sprawdzane w warstwie biznesowej a nie tylko prezentacji. Nie jest tak? :thinking:

Reply:
> No i warto dodać, że treść błędu to coś w stylu "pole xxx nie może być nullem", więc szczegóły implementacji lecą na front.
> W takim przypadku dodaje się jakieś mappery, które przerabiają ten null pointer na jakiś error http?

Czy czasem nie jest tak, że każdy exception który nie ma ustawionego explicite kodu HTTP powoduje 500? Czy jego treść leci niezależnie od trybu dev?

Reply:
no okej, niech i tak będzie

leci 500, ale jest error message jak wyżej

Reply:
I o ile błędy mówiące o tym, że podany user nie jest w projekcie mają sens, to błąd, że jakieś pole nie jest ustawione już nie (bo to już jest ewidentnie wina implementującego ;))

Reply:
> I o ile błędy mówiące o tym, że **(1)** podany user nie jest w projekcie mają sens, to błąd, że **(2)** jakieś pole nie jest ustawione już nie (bo to już jest ewidentnie wina implementującego ;))

1. Wtedy powinien lecieć błąd 4xx z message.
2. Wtedy powinien lecieć błąd 5xx bez message (z message w trybie dev).

Tak mi się wydaje przynajmniej.

Reply:
ach, to nie spodziewałem się, że wysyłanie message zależy od profilu xd (oczywiście, że siedzę na dev)

W takim razie plus dla Ciebie, za fajny ficzer

Reply:
Ja opisywałem tylko jak to _powinno_ wyglądać. Nie mam pojęcia jak jest w rzeczywistości :rofl: